File: spring-backend/src/main/java/com/creditwise/entity/LoanProduct.java
package com.creditwise.entity;

import jakarta.persistence.*;
import lombok.*;

/**
 * Loan Product Entity
 * ===================
 * Represents loan products from LendingClub and other lenders.
 * Contains real APR rates based on credit score grades.
 */
@Entity
@Table(name = "loan_products")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class LoanProduct {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "product_id", unique = true, nullable = false)
    private String productId;

    // =========================================================================
    // Product Information
    // =========================================================================
    
    @Column(name = "product_name", nullable = false)
    private String productName;

    @Column(name = "lender_name", nullable = false)
    private String lenderName;

    @Column(name = "loan_type", nullable = false)
    private String loanType;  // personal, debt_consolidation, home, car, education

    @Column(nullable = false)
    private String grade;  // A, B, C, D, E

    @Column(length = 500)
    private String description;

    // =========================================================================
    // Credit Requirements
    // =========================================================================
    
    @Column(name = "min_credit_score", nullable = false)
    private Integer minCreditScore;

    @Column(name = "max_credit_score", nullable = false)
    private Integer maxCreditScore;

    // =========================================================================
    // Loan Terms
    // =========================================================================
    
    @Column(name = "min_apr", nullable = false)
    private Double minApr;

    @Column(name = "max_apr", nullable = false)
    private Double maxApr;

    @Column(name = "min_amount", nullable = false)
    private Double minAmount;

    @Column(name = "max_amount", nullable = false)
    private Double maxAmount;

    @Column(name = "terms_months", nullable = false)
    private String termsMonths;  // Comma-separated: "36,48,60"

    @Column(name = "origination_fee_pct")
    @Builder.Default
    private Double originationFeePct = 0.0;

    @Column(name = "is_active")
    @Builder.Default
    private Boolean isActive = true;
}

File: backend/app/models/loan_product.py
"""
Loan Product Database Model
===========================
SQLAlchemy model for loan products based on LendingClub data.
"""

from sqlalchemy import Column, Integer, String, Float, Boolean
from ..core.database import Base


class LoanProduct(Base):
    """
    Loan products available to users based on their credit profile.
    
    Data sourced from LendingClub, SoFi, Prosper, Upstart, Marcus, and Discover
    public rate sheets and SEC filings.
    
    Table: loan_products
    """
    
    __tablename__ = "loan_products"
    
    # ==========================================================================
    # Primary Key
    # ==========================================================================
    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    product_id = Column(String(50), unique=True, nullable=False, index=True)
    
    # ==========================================================================
    # Product Information
    # ==========================================================================
    product_name = Column(String(100), nullable=False)
    lender_name = Column(String(100), nullable=False)
    loan_type = Column(String(50), nullable=False)  # personal, debt_consolidation, home, car, education, business
    grade = Column(String(5), nullable=False)  # A, B, C, D, E, F, G
    description = Column(String(500), nullable=True)
    
    # ==========================================================================
    # Credit Requirements
    # ==========================================================================
    min_credit_score = Column(Integer, nullable=False)
    max_credit_score = Column(Integer, nullable=False)
    
    # ==========================================================================
    # Loan Terms
    # ==========================================================================
    min_apr = Column(Float, nullable=False)  # Minimum APR %
    max_apr = Column(Float, nullable=False)  # Maximum APR %
    min_amount = Column(Float, nullable=False)  # Minimum loan amount
    max_amount = Column(Float, nullable=False)  # Maximum loan amount
    terms_months = Column(String(50), nullable=False)  # Comma-separated: "36,48,60"
    origination_fee_pct = Column(Float, default=0.0)  # Origination fee %
    
    # ==========================================================================
    # Status
    # ==========================================================================
    is_active = Column(Boolean, default=True)
    
    def __repr__(self):
        return f"<LoanProduct(id={self.product_id}, name='{self.product_name}', apr={self.min_apr}-{self.max_apr}%)>"
    
    def to_dict(self):
        return {
            "id": self.id,
            "product_id": self.product_id,
            "product_name": self.product_name,
            "lender_name": self.lender_name,
            "loan_type": self.loan_type,
            "grade": self.grade,
            "description": self.description,
            "min_credit_score": self.min_credit_score,
            "max_credit_score": self.max_credit_score,
            "min_apr": self.min_apr,
            "max_apr": self.max_apr,
            "min_amount": self.min_amount,
            "max_amount": self.max_amount,
            "terms_months": self.terms_months,
            "origination_fee_pct": self.origination_fee_pct,
            "is_active": self.is_active,
        }

